go/extractor/diagnostics/diagnostics.go

```diff
@@ -0,0 +1,184 @@
+		} else {
+			optLoc = &locationStruct{file, startLine, startColumn, endLine, endColumn}
+		}
+
+		timestamp := time.Now().UTC().Format("2006-01-02T15:04:05.000") + "Z"
+
+		d := diagnostic{
+			timestamp,
+			sourceStruct{sourceid, sourcename, "go"},
+			markdownMessage,
+			string(severity),
+			internal,
+			visibilityStruct{visibilitySP, visibilityCST, visibilityT},
+			optLoc,
+		}
+
+		if diagnosticsEmitted == diagnosticsLimit {
+			d = diagnostic{
+				timestamp,
+				sourceStruct{"go/autobuilder/diagnostic-limit-hit", "Some diagnostics were dropped", "go"},
+				fmt.Sprintf("The number of diagnostics exceeded the limit (%d); the remainder were dropped.", diagnosticsLimit),
+				string(severityWarning),
+				false,
+				visibilityStruct{true, true, true},
+				nil,
+			}
+		}
+
+		content, err := json.Marshal(d)
+		if err != nil {
+			log.Println(err)
+		}
+
+		targetFile, err := os.CreateTemp(diagnosticDir, "go-extractor.*.jsonl")
+		if err != nil {
+			log.Println("Failed to create temporary file for diagnostic: ")
+			log.Println(err)
+		}
+		defer func() {
+			if err := targetFile.Close(); err != nil {
+				log.Println("Failed to close diagnostic file:")
+				log.Println(err)
+			}
+		}()
+
+		_, err = targetFile.Write(content)
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+}
+
+func EmitPackageDifferentOSArchitecture(pkgPath string) {
+	emitDiagnostic(
+		"go/autobuilder/package-different-os-architecture",
+		"Package "+pkgPath+" is intended for a different OS or architecture",
+		"Make sure the `GOOS` and `GOARCH` [environment variables are correctly set](https://docs.github.com/en/actions/learn-github-actions/variables#defining-environment-variables-for-a-single-workflow). Alternatively, [change your OS and architecture](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#using-a-github-hosted-runner).",
+		severityWarning, false,
+		true, true, true,
+		"", 0, 0, 0, 0,
+	)
+}
+
+const maxNumPkgPaths = 5
+
+func EmitCannotFindPackages(pkgPaths []string) {
+	numPkgPaths := len(pkgPaths)
+
+	ending := "s"
+	if numPkgPaths == 1 {
+		ending = ""
+	}
+
+	numPrinted := numPkgPaths
+	truncated := false
+	if numPrinted > maxNumPkgPaths {
+		numPrinted = maxNumPkgPaths
+		truncated = true
+	}
+
+	secondLine := "`" + strings.Join(pkgPaths[0:numPrinted], "`, `") + "`"
+	if truncated {
+		secondLine += fmt.Sprintf(" and %d more", numPkgPaths-maxNumPkgPaths)
+	}
+
+	emitDiagnostic(
+		"go/autobuilder/package-not-found",
+		fmt.Sprintf("%d package%s could not be found", numPkgPaths, ending),
+		"The following packages could not be found. Check that the paths are correct and make sure any private packages can be accessed. If any of the packages are present in the repository then you may need a [custom build command](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/configuring-the-codeql-workflow-for-compiled-languages).\n\n"+secondLine,
```

Comment:
Should this be re-arranged to include the list of packages in the middle and then the user action at the end?

Reply:
I slightly prefer it this way, but I agree the guidance indicates I should change it. I will double-check and then probably change it.

Reply:
I've changed it.